caddytls/ctmonitor.go

```diff
@@ -0,0 +1,308 @@
+		return
+	}
+	for {
+		namesToLookUp, caddyCerts := getCaddyCerts()
+		if len(namesToLookUp) == 0 {
+			log.Printf("[ERROR]Could not retrieve DNS names from Caddy Certificate, " +
+				"make sure that you are serving on port 80 & 443. " +
+				"Terminating monitorCerts.\n")
+		}
+		startingIndex, err := getLatestIndex(filePath)
+		if err != nil {
+			log.Printf("Error %v while getting starting index, starting at 0", err)
+			break
+		}
+		config.Query = certSpotterAPIBase
+		config.Index = startingIndex
+		fetchedCerts, pause, err := lookUpNames(namesToLookUp, config)
+
+		// CompareCerts runs even if there was an error.  If the error occurred after
+		// retrieving all of the certs, we can still compare them before quitting.
+		CompareCerts(caddyCerts, fetchedCerts)
+		if err != nil {
+			log.Printf("[ERROR] lookUpNames encountered an error %v, terminating ctmonitor.", err)
+			break
+		}
+		time.Sleep(time.Duration(pause) * time.Second)
+	}
+}
+
+func prepQuery(domainName string, config QueryConfig) (query string) {
+	v := url.Values{}
+	v.Set("domain", domainName)
+	if config.CTConfig.Include_wildcards {
+		v.Set("match_wildcards", "true")
+	}
+	if config.CTConfig.Include_subdomains {
+		v.Set("include_subdomains", "true")
+	}
+	v.Set("after", strconv.Itoa(config.Index))
+	v.Add("expand", "dns_names")
+	v.Add("expand", "issuer")
+	v.Add("expand", "cert")
+	encodedValue := v.Encode()
+	query = config.Query + "?" + encodedValue
+	return query
+}
+
+func putLatestID(currentID int, fileName string) error {
+	writeValue := strconv.Itoa(currentID)
+	return ioutil.WriteFile(fileName, []byte(writeValue), 0600)
+}
+
+func queryDomainName(domainName string, config QueryConfig, biggestID *int, retrievedCerts map[string]string) (retryAfter string, err error) {
+	var querySize int
+
+	for ok := true; ok; ok = querySize > 0 {
+		querySize, config.Index, retryAfter, err = getCertSpotterCerts(domainName, config, biggestID, retrievedCerts)
+		if err != nil {
+			return retryAfter, fmt.Errorf("error: %v while getting certs from certSpotter", err) // Finish error checking.
+		}
+	}
+	return
+}
+
+func startMonitoring(eventType caddy.EventName, eventInfo interface{}) error {
+	go monitorCerts()
+	return nil
+}
+
+func getCertSpotterCerts(domainName string, config QueryConfig, biggestID *int, retrievedCerts map[string]string) (numOfIssuanceObjects int, issuanceID int, retryAfter string, err error) {
+	// issuanceObjects consists of the issuanceObjects returned from CertSpotter.
+	issuanceID = config.Index
+	var (
+		issuanceObjects []CertSpotterResponse
+		certQuery       string
+	)
+	certQuery = prepQuery(domainName, config)
+
+	response, err := http.Get(certQuery)
+	if err != nil {
+		log.Printf("https get request failed on input %s \nError: %v", prepQuery(domainName, config), err)
+	}
+	defer response.Body.Close()
+
+	err = json.NewDecoder(response.Body).Decode(&issuanceObjects) // handle error
+	if err != nil {
+		retryAfter = "3600"
```

Comment:
Still prefer the Retry-After header even if the response body is malformed.

Reply:
I think that this value is only returned if the issuance endpoint returns an empty array.
From their page: 

Polling For New Issuances

When the issuances endpoint returns an empty array, it includes a Retry-After HTTP header containing a number of seconds. After the specified time has elapsed, you may retry the request. If Cert Spotter has discovered new issuances matching your query, they will be returned. You can then request subsequent pages until another empty response is returned.

Reply:
This is what I'm doing now, to make sure that there is some waiting no matter what happens.
if retryAfter == "" {
     retryAfter = defaultWaitTime
}